src/org/nfrac/comportex/cells.cljc

```diff
@@ -735,58 +737,97 @@
                                                     (:distal-bits distal-state)
                                                     (:distal-exc distal-state)
                                                     (:seg-learn-threshold spec)
                                                     (:depth spec))
                             (merge-with + tp-exc))
           ;; find active and winner cells in the columns
+          pc (:pred-cells distal-state)
+          depth (:depth spec)
+          prior-col-winners (:col-winners state)
           [rng* rng] (random/split rng)
           {ac :active-cells
-           lc :winner-cells
-           b-cols :burst-cols
+           col-winners :col-winners
+           burst-cols :burst-cols
            stable-ac :stable-active-cells}
-          (select-active-cells a-cols rel-cell-exc (:pred-cells distal-state) spec rng*)
+          (select-active-cells a-cols rel-cell-exc
+                               ;; definition of bursting for a column
+                               (fn [col win-cell col-ac]
+                                 (if (and (not newly-engaged?)
+                                          (= win-cell (prior-col-winners col)))
+                                   ;; for continuing temporal pooling
+                                   (== depth (count col-ac))
+                                   ;; otherwise: for discrete transitions
+                                   (not (or (pc win-cell) (tp-exc win-cell)))))
+                               (:col-winners state) ;; keep winners stable
+                               spec rng*)
+          ;; learning cells are the winning cells, but excluding any
+          ;; continuing winners when temporal pooling
+          old-winners (vals (:col-winners state))
+          new-winners (vals col-winners)
+          learning (if newly-engaged? ;; always true at first level
+                     new-winners
+                     (remove (set old-winners) new-winners))
           ;; update continuing TP activation
-          lc-stable-exc (for [cell-id lc
-                              :let [[col ci] cell-id
-                                    exc (stable-col-exc [col 0])]
-                              :when exc]
-                          [cell-id exc])
-          next-tp-exc (-> (apply-excitation tp-exc
-                                            lc-stable-exc
-                                            (:temporal-pooling-max-exc spec)
-                                            (:temporal-pooling-amp spec))
-                          ;; clear TP excitation for inhibited cells
-                          (select-keys lc))
-          depth (:depth spec)]
+          next-tp-exc (if higher-level?
+                        (let [new-ac (if newly-engaged?
+                                       ac
+                                       (set/difference ac (:active-cells state)))]
+                          (into tp-exc
```

Comment:
Yes, that's right. But I think it makes more sense if you consider the proximal excitation together with the temporal pooling excitation.
* if the layer is newly engaged, any existing TP is cleared, columns are selected using all proximal input, and the activated cells get the full TP amount.
* if the layer is continuing engaged, columns are selected using all proximal input together with the existing TP. Because we select a larger number of columns each step, they will be the existing TP ones plus some new ones. The new ones (cells) get the full TP amount.
* if the layer is not engaged, no proximal input comes through, with the exception of any columns having a high match. So TP columns will just continue. The idea was to carry forward context (including transition predictions) rather than forgetting it. But I just realised that it resets the activation level so it won't carry forward unchanged, d'oh.

A thought, inspired by [this](https://github.com/numenta/nupic.research/wiki/Behavior-and-Pooling-Requirements), "Novel input should appear stable": when a layer is newly **not** engaged, i.e. something novel appears, we could reset TP and start pooling again! That way we build up a representation of the novel thing, until we start to recognise something and reset (transition) again. But if not engaged then we don't learn proximally.

Whether TP amounts should always decay, or only when the layer is not engaged, I'm not sure. The former would bias the representation towards more recent steps in the pooled sequence. Since we allocate more columns for each step this seems like it should not be necessary, but it would be once we fill up to the maximum density/sparsity.